<a href="https://colab.research.google.com/github/JenniferZeng11/world-covid-vaccinations/blob/main/worldcovidvaccination.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

**Describe why you've picked the dataset**<br>
To fight against the Covid-19 pandemic, researchers and scientists did their best to come up with vaccinations. I am interested in exploring more about the covid vaccination population around the world, with the skills we learned in this class, pandas, plotyly and Tableau. Questions to be explored include vaccinations number by country, which country has the most people vaccinated daily, and what about in percent? <br>

**About the data**<br>
The current exploration utilized the COVID-19 World Vaccination Progress data from Kaggle with version 114 updated by May 6.  


In [1]:
if 'google.colab' in str(get_ipython()):
    !pip install plotly==4.14.3

In [2]:
import pandas as pd
import numpy as np
import plotly
import plotly.express as px
import plotly.graph_objects as go


import plotly.io as pio

In [3]:
df = pd.read_csv('https://github.com/JenniferZeng11/world-covid-vaccinations/blob/main/country_vaccinations.csv?raw=true')

In [4]:
df.shape

(15505, 15)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15505 entries, 0 to 15504
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   country                              15505 non-null  object 
 1   iso_code                             15505 non-null  object 
 2   date                                 15505 non-null  object 
 3   total_vaccinations                   9325 non-null   float64
 4   people_vaccinated                    8649 non-null   float64
 5   people_fully_vaccinated              6425 non-null   float64
 6   daily_vaccinations_raw               7830 non-null   float64
 7   daily_vaccinations                   15307 non-null  float64
 8   total_vaccinations_per_hundred       9325 non-null   float64
 9   people_vaccinated_per_hundred        8649 non-null   float64
 10  people_fully_vaccinated_per_hundred  6425 non-null   float64
 11  daily_vaccinations_per_milli

In [6]:
df.head()

country  ...                                     source_website
0  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
1  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
2  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
3  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
4  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...

[5 rows x 15 columns]

In [7]:
df.tail()

country  ...                                     source_website
15500  Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1389675251...
15501  Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1389675251...
15502  Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1389675251...
15503  Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1389675251...
15504  Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1389675251...

[5 rows x 15 columns]

In [8]:
df.head(10)

country  ...                                     source_website
0  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
1  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
2  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
3  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
4  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
5  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
6  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
7  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
8  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
9  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...

[10 rows x 15 columns]

# Total Vacinnations by country

**Total Vacinnations by country Table**

In [9]:
df_country_total_va = df.groupby('country')['total_vaccinations'].max()
df_country_total_va = df_country_total_va.reset_index()
df_country_total_va

country  total_vaccinations
0    Afghanistan            240000.0
1        Albania            556584.0
2        Algeria             75000.0
3        Andorra             28881.0
4         Angola            456349.0
..           ...                 ...
190    Venezuela            250000.0
191      Vietnam            585539.0
192        Wales           2656272.0
193       Zambia             58982.0
194     Zimbabwe            559777.0

[195 rows x 2 columns]

**Total Vaccinations by Countries Histogram**

In [10]:
fig = px.histogram(df_country_total_va,
             x='total_vaccinations',
             title='Total Vaccinations by Countries')
fig.show()

The 'Total Vaccinations by Countries' histogram show the total vaccination distribution. Most of the countires were in the range of 0-9.9 million population for total vaccinations. 

**Total Vaccinations by Top 10 Countries**

In [11]:
df_country_total_va.nlargest(10, ['total_vaccinations'])

country  total_vaccinations
34            China         289627000.0
187   United States         249566820.0
78            India         159931238.0
186  United Kingdom          50682567.0
25           Brazil          45671672.0
51          England          42408492.0
64          Germany          32596999.0
181          Turkey          24103883.0
60           France          23404537.0
85            Italy          22112154.0

In [12]:
fig = px.bar(
    df_country_total_va.nlargest(10, ['total_vaccinations']),
    x='total_vaccinations',
    y='country',
    color = 'country',
    title='Total Vaccinations by Top 10 Countries',
    height=600
)

fig.update_yaxes(categoryorder='total ascending')

fig.show()

This histogram above 'Total Vaccinations by Top 10 Countries' showed the top 10 countries' total vaccination numbers, with the highest in the range from 200 to 290 million people. The table also indicated that China (289627000 people) was the country with the most vaccinations, followed by United States (249566820 people).

**Total Vaccinations by Bottom 10 Countries**

In [13]:
df_country_total_va.nsmallest(10, ['total_vaccinations'])

country  total_vaccinations
29                       Cameroon               400.0
8                         Armenia               565.0
121                         Nauru               700.0
99                          Libya               750.0
166                   South Sudan               947.0
126                         Niger              1366.0
43   Democratic Republic of Congo              1700.0
116                    Montserrat              1909.0
31                     Cape Verde              2184.0
173                         Syria              2500.0

In [14]:
fig = px.bar(
    df_country_total_va.nsmallest(10, ['total_vaccinations']),
    x='total_vaccinations',
    y='country',
    color = 'country',
    title='Total Vaccinations by Bottom 10 Countries',
    height=600
)

fig.update_yaxes(categoryorder='total descending')

fig.show()

This histogram above 'Total Vaccinations by Bottom 10 Countries' showed the bottom 10 countries' total vaccination numbers, with the lowest in the range from 0 to 990 people. The number pointed out that Cameroon	was the country with the lowest number of total vaccination with 400 people.

# Fully Vacinnations by Country

**Fully Vacinnations by Country Table**

In [15]:
df_country_fully_va = df.groupby('country')['people_fully_vaccinated'].max()
df_country_fully_va = df_country_fully_va.reset_index()
df_country_fully_va

country  people_fully_vaccinated
0    Afghanistan                      NaN
1        Albania                    655.0
2        Algeria                      NaN
3        Andorra                   4699.0
4         Angola                      NaN
..           ...                      ...
190    Venezuela                      NaN
191      Vietnam                      NaN
192        Wales                 783120.0
193       Zambia                      NaN
194     Zimbabwe                 107586.0

[195 rows x 2 columns]

**cleaning data: drop the rows with NaN data**

In [16]:
df_country_fully_va  = df_country_fully_va .dropna(axis=0,how='any')
df_country_fully_va = df_country_fully_va .reset_index(drop = True)
df_country_fully_va 

country  people_fully_vaccinated
0           Albania                    655.0
1           Andorra                   4699.0
2          Anguilla                    783.0
3         Argentina                1130337.0
4             Aruba                  23824.0
..              ...                      ...
124  United Kingdom               15887493.0
125   United States              107346533.0
126         Uruguay                 733467.0
127           Wales                 783120.0
128        Zimbabwe                 107586.0

[129 rows x 2 columns]

**Fully Vaccinations by Countries Boxplot**

In [17]:
fig = px.box(df_country_fully_va,
             x='people_fully_vaccinated',
             orientation='h',
             title='Fully Vaccinations by Countries')
fig.show()

In [18]:
#(excluding outliers for better view)
fig = px.box(df_country_fully_va,
             x='people_fully_vaccinated',
             orientation='h',
             points=False,
             title='Fully Vaccinations by Countries (without outliers)')
fig.show()

# Maximun Daily Vaccinations by Country Comparison

**Maximun Daily Vaccinations by Country Table**

In [19]:
df_country_daily_va = df.groupby('country')['daily_vaccinations'].max()
df_country_daily_va = df_country_daily_va.reset_index()
df_country_daily_va

country  daily_vaccinations
0    Afghanistan              8000.0
1        Albania             17565.0
2        Algeria              3748.0
3        Andorra              1182.0
4         Angola             18751.0
..           ...                 ...
190    Venezuela             15200.0
191      Vietnam             53975.0
192        Wales             33151.0
193       Zambia              5295.0
194     Zimbabwe             21171.0

[195 rows x 2 columns]

**cleaning data: drop the rows with NaN data**

In [20]:
df_country_daily_va  = df_country_daily_va .dropna(axis=0,how='any')
df_country_daily_va = df_country_daily_va .reset_index(drop = True)
df_country_daily_va

country  daily_vaccinations
0    Afghanistan              8000.0
1        Albania             17565.0
2        Algeria              3748.0
3        Andorra              1182.0
4         Angola             18751.0
..           ...                 ...
184    Venezuela             15200.0
185      Vietnam             53975.0
186        Wales             33151.0
187       Zambia              5295.0
188     Zimbabwe             21171.0

[189 rows x 2 columns]

In [21]:
fig = px.density_heatmap(
    df_country_daily_va,
    x = 'daily_vaccinations',
    y = 'country',
    title = 'daily vaccinations heatmap',
    height = 600
)

fig.show()

**Top 10 Country of Daily Vaccinations**


In [22]:
df_country_daily_va.nlargest(10, ['daily_vaccinations'])

country  daily_vaccinations
33            China           7205286.0
75            India           3659357.0
181   United States           3384387.0
24           Brazil           1135847.0
61          Germany            672360.0
180  United Kingdom            602266.0
48          England            507875.0
76        Indonesia            449308.0
82            Italy            447864.0
175          Turkey            435596.0

In [23]:
fig = px.treemap(
    df_country_daily_va.nlargest(10, ['daily_vaccinations']),
    path=['country'],
    title='Top 10 Daily vaccinations',
    values='daily_vaccinations',
    height=700
)

fig.show()

# Most up to date data

**Checking the most recent date of the data**

In [24]:
df['date'].max()

'2021-05-05'

**Daily Vaccinations United States on May 5th**

In [25]:
df_may_5 = df[df['date'] == '2021-05-05']
df_may_5

country  ...                                     source_website
175                 Albania  ...  https://twitter.com/gmanastirliu/status/138999...
622               Argentina  ...  http://datos.salud.gob.ar/dataset/vacunas-cont...
661                   Aruba  ...                          https://www.government.aw
741               Australia  ...              https://covidlive.com.au/vaccinations
871                 Austria  ...  https://info.gesundheitsministerium.gv.at/open...
...                     ...  ...                                                ...
14694               Ukraine  ...                https://vaccination.covid19.gov.ua/
14815  United Arab Emirates  ...                     http://covid19.ncema.gov.ae/en
15074         United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
15142               Uruguay  ...            https://monitor.uruguaysevacuna.gub.uy/
15428                Zambia  ...  https://www.facebook.com/mohzambia/photos/a.13...

[62 rows x 15 columns]

**Cleaning May 5th Data**

In [26]:
df_may_5  = df_may_5.dropna(axis=0,how='any')
df_may_5 = df_may_5.reset_index(drop = True)
df_may_5

country  ...                                     source_website
0             Argentina  ...  http://datos.salud.gob.ar/dataset/vacunas-cont...
1                 Aruba  ...                          https://www.government.aw
2               Austria  ...  https://info.gesundheitsministerium.gv.at/open...
3            Azerbaijan  ...                         https://koronavirusinfo.az
4               Bahrain  ...                     https://healthalert.gov.bh/en/
5            Bangladesh  ...  https://dghs.gov.bd/images/docs/vpr/Covid-19-V...
6               Belgium  ...                  https://epistat.wiv-isp.be/covid/
7                Brazil  ...  https://qsprod.saude.gov.br/extensions/DEMAS_C...
8              Bulgaria  ...               https://coronavirus.bg/bg/statistika
9              Cambodia  ...  http://www.freshnewsasia.com/index.php/en/loca...
10               Canada  ...  https://covid19tracker.ca/vaccinationtracker.html
11              Croatia  ...  https://www.index.hr/vijesti/clanak/sluzbeno-u...
12              Curacao  ...                                 https://bakuna.cw/
13              Czechia  ...       https://onemocneni-aktualne.mzcr.cz/covid-19
14              Estonia  ...                       https://opendata.digilugu.ee
15              Finland  ...  https://sampo.thl.fi/pivot/prod/en/vaccreg/cov...
16              Germany  ...                          https://impfdashboard.de/
17               Greece  ...      https://www.data.gov.gr/datasets/mdg_emvolio/
18            Hong Kong  ...       https://www.covidvaccine.gov.hk/en/dashboard
19              Hungary  ...  https://koronavirus.gov.hu/cikkek/4-millio-153...
20                India  ...                    https://dashboard.cowin.gov.in/
21          Isle of Man  ...  https://covid19.gov.im/general-information/cov...
22               Israel  ...  https://datadashboard.health.gov.il/COVID-19/g...
23                Italy  ...  https://raw.githubusercontent.com/italia/covid...
24           Kazakhstan  ...                    https://www.coronavirus2020.kz/
25               Latvia  ...  https://data.gov.lv/dati/eng/dataset/covid19-v...
26              Lebanon  ...   https://impact.cib.gov.lb/home/dashboard/vaccine
27            Lithuania  ...  https://ls-osp-sdg.maps.arcgis.com/apps/opsdas...
28                Macao  ...            https://www.gov.mo/zh-hant/news/376099/
29              Moldova  ...                          https://vaccinare.gov.md/
30             Mongolia  ...                                   https://ikon.mn/
31           Montenegro  ...                       https://www.covidodgovor.me/
32               Norway  ...  https://www.fhi.no/sv/vaksine/koronavaksinasjo...
33               Panama  ...  https://twitter.com/MINSAPma/status/1390085715...
34             Portugal  ...          https://github.com/dssg-pt/covid19pt-data
35                Qatar  ...  https://covid19.moph.gov.qa/EN/Pages/Vaccinati...
36              Romania  ...                               https://datelazi.ro/
37               Russia  ...            https://gogov.ru/articles/covid-v-stats
38             Slovakia  ...  https://github.com/Institut-Zdravotnych-Analyz...
39             Slovenia  ...            https://covid-19.sledilnik.org/en/stats
40          South Korea  ...                             http://ncv.kdca.go.kr/
41            Sri Lanka  ...  https://www.epid.gov.lk/web/images/pdf/corona_...
42               Sweden  ...  https://www.folkhalsomyndigheten.se/smittskydd...
43             Thailand  ...  https://ddc.moph.go.th/uploads/ckeditor2//file...
44  Trinidad and Tobago  ...  https://experience.arcgis.com/experience/59226...
45               Turkey  ...                  https://covid19asi.saglik.gov.tr/
46              Ukraine  ...                https://vaccination.covid19.gov.ua/
47        United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
48              Uruguay  ...            https://monitor.uruguaysevacuna.gub.uy/

[49 rows x 15 c

**Top 10 Countries of People Fully Vaccinated by May 5th Bar Chart**

In [27]:
fig = px.bar(
    df_may_5.nlargest(10, ['people_fully_vaccinated']),
    x='people_fully_vaccinated',
    y='country',
    color='vaccines',
    template='plotly_dark',
    title='Top 10 Countries of People Fully Vaccinated by May 5th',
    height=600
)

fig.show()

**Vaccines Type Pie Chart**

In [28]:
fig = px.pie(
    df_may_5,
    names='vaccines',
    title='Vaccines Type',
    width=800,
    height=700
)

fig.show()

Note: The vaccine column contains all the vaccine information for one country; the pie chart could not clearly represent the different vaccine breakdown. 

# United States Vaccination 

**United States Data Table**

In [29]:
df_us = df[df['country'] == 'United States']
df_us

country  ...                                     source_website
14938  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
14939  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
14940  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
14941  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
14942  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
...              ...  ...                                                ...
15070  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
15071  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
15072  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
15073  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
15074  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...

[137 rows x 15 columns]

**Cleaned data**

In [30]:
df_us  = df_us.dropna(axis=0,how='any')
df_us = df_us.reset_index(drop = True)
df_us

country  ...                                     source_website
0    United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
1    United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
2    United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
3    United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
4    United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
..             ...  ...                                                ...
101  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
102  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
103  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
104  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...
105  United States  ...  https://covid.cdc.gov/covid-data-tracker/#vacc...

[106 rows x 15 columns]

**Daily Vaccination in US Scatter Plot**

In [31]:
fig = px.scatter(
    df_us, 
    x = 'date', 
    y = 'daily_vaccinations', 
    title = 'Daily Vaccination in US Scatter Plot')
fig.show()

**Daily Vaccination in United States**

In [32]:
fig = px.line(
    df_us, 
    x = 'date', 
    y = 'people_fully_vaccinated', 
    title = 'Daily Vaccination in United States')
fig.show()